proc_macros/src/rule_tests.rs

```diff
@@ -3,20 +3,100 @@ use quote::{format_ident, quote, ToTokens};
 use syn::{
     braced, bracketed,
     parse::{Parse, ParseStream},
-    parse_macro_input, Expr, ExprArray, Ident, Token,
+    parse_macro_input, token, Expr, ExprArray, Ident, Token,
 };
 
+struct RuleOptions {
+    options: Expr,
+}
+
+impl Parse for RuleOptions {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let options: Expr = input.parse()?;
+        Ok(Self { options })
+    }
+}
+
+impl ToTokens for RuleOptions {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let options = &self.options;
+        quote! {
+            #options.into()
+        }
+        .to_tokens(tokens)
+    }
+}
+
+struct ValidRuleTestSpec {
+    code: Expr,
+    options: Option<RuleOptions>,
+}
+
+impl Parse for ValidRuleTestSpec {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut code: Option<Expr> = Default::default();
+        let mut options: Option<RuleOptions> = Default::default();
+        if input.peek(token::Brace) {
```

Comment:
Mimic ESLint and allow specifying either just a simple string (corresponding to the `code` key) or a "key-value blob" (if you also need to supply eg `options`) (for the `valid` "specs" in rule tests)